### DISCUSSION TITLE
change prompt within trial

### DISCUSSION THREAD

**Discussion by petyaracz** [jspsych/jsPsych#2645] [Q&A] [answered]
I'd like to set up a button press trial. 

The participant is told that another player is doing the task concurrently (this is basically Von Ahn's ESP paradigm). The other player's answers will be scripted.

I'd like to have the other player "think" for a few seconds in each trial. This should not affect trial structure.

I could do this like so. The problem here is that this counts as two trials. If the participant picks an answer _before_ the other player stopped "thinking", they will have to pick an answer again to proceed.

I'd like to change the prompt from "other player thinking" to "other player picked" within trial, so that the trial only registers one button press. This might be impossible.

```
<!DOCTYPE html>
<html>
  <head>
    <title>Fake dyad experiment</title>
    <script src="https://unpkg.com/jspsych@7.0.0"></script>
    <script src="https://unpkg.com/@jspsych/plugin-html-button-response@1.0.0"></script>
    <link href="https://unpkg.com/jspsych@7.0.0/css/jspsych.css" rel="stylesheet" type="text/css" />
    <meta charset="UTF-8">
  </head>
  <body></body>
  <script>
    const jsPsych = initJsPsych();

    var trial = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'The name of a popular red fruit is:',
      choices: ['apple','carburator'],
      timeline:[
          {
            prompt:'<p>Player 2 is thinking. &#129300;</p>',
            trial_duration:5000
          },
          {
            prompt:'<p>Player 2 made a pick! &#128526;</p>'
          }
        ]
      };

    var final = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'Thanks, bye.',
      choices: ['Ok']
    };

    jsPsych.run([trial,final]);
  </script>
</html>
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @petyaracz, here's how I would do this. (Standard disclaimer: this may not be the best/easiest way)

```html
<!DOCTYPE html>
<html>
  <head>
    <title>Fake dyad experiment</title>
    <script src="https://unpkg.com/jspsych@7.0.0"></script>
    <script src="https://unpkg.com/@jspsych/plugin-html-button-response@1.0.0"></script>
    <link href="https://unpkg.com/jspsych@7.0.0/css/jspsych.css" rel="stylesheet" type="text/css" />
    <meta charset="UTF-8">
  </head>
  <body></body>
  <script>
    const jsPsych = initJsPsych();

    var trial = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'The name of a popular red fruit is:',
      choices: ['apple','carburator'],
      prompt:'<p>Player 2 is thinking. &#129300;</p>',
      on_load: function() {
        // for some reason the prompt element doesn't have an ID, 
        // but you can find it by getting the last child element in the jspsych-content div
        const prompt = document.getElementById('jspsych-content').lastElementChild;
        // set up a timer to change the prompt text after a given delay
        jsPsych.pluginAPI.setTimeout(function() { 
            prompt.innerHTML = '<p>Player 2 made a pick! &#128526;</p>'
        }, 5000)
      },
      on_finish: function() {
        // clear the timeout at the end of the trial, in case a response was made and the trial ends
        // before the timeout duration is reached
        jsPsych.pluginAPI.clearAllTimeouts();
      }
    };

    var final = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'Thanks, bye.',
      choices: ['Ok']
    };

    jsPsych.run([trial,final]);
  </script>
</html>
```
Does that work for you? 
You could also randomize the delay inside the `on_load` function, or use timeline variables to run a set of trials with specific delays.

**Reply by petyaracz** [reply]
Thank you. This is definite improvement on my solution. The problem remains that it ends the trial when the player picked, irrespective of whether the timer ran out.

My flow should have the following possibilities:

1. Player picks before timer runs out. Player's response is stored for the trial but trial doesn't end before timer runs out.
2. Player picks after timer runs out. Prompt tells player that timer ran out ("player 2 finished") but doesn't end the trial.

Currently, both solutions cover (2) but not (1). (Also my solution doesn't work!) 

(1) is easy to do separately (you end the trial on a hard time limit instead of a choice). (2) is covered by @becky-gilbert 's solution. 

Is it possible to combine the two? 

Maybe tell the trial `on_finish` to NOT end trial on choice if timer hasn't ran out? I wrote this up in pseudocode (I don't really know which variables to refer to, here, or if this works at all):

```
var trial = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'The name of a popular red fruit is:',
      choices: ['apple','carburator'],
      prompt:'<p>Player 2 is thinking. &#129300;</p>',
      on_load: function() {
        // for some reason the prompt element doesn't have an ID, 
        // but you can find it by getting the last child element in the jspsych-content div
        const prompt = document.getElementById('jspsych-content').lastElementChild;
        // set up a timer to change the prompt text after a given delay
        jsPsych.pluginAPI.setTimeout(function() { 
            prompt.innerHTML = '<p>Player 2 made a pick! &#128526;</p>'
        }, 5000)
      },
      on_finish: function() {
        // where timeout is jsPsych counting back from 5000:
        if (player picked something & timeout != 0){
          wait anyway
        }
      }
    };
```

**Reply by jodeleeuw** [reply]
What about using `response_ends_trial: false` and `trial_duration` together with this solution? That seems like it would solve (1).

**Reply by petyaracz** [reply]
Can I tell jspsych to end the trial on response IF trial duration is longer than the set timer (that is, player 2 is done "thinking")?

Something like

```
    var player_2_timer = 5000;

    var trial = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'The name of a popular red fruit is:',
      choices: ['apple','carburator'],
      prompt:'<p>Player 2 is thinking. &#129300;</p>',
      on_load: function() {
        // for some reason the prompt element doesn't have an ID, 
        // but you can find it by getting the last child element in the jspsych-content div
        const prompt = document.getElementById('jspsych-content').lastElementChild;
        // set up a timer to change the prompt text after a given delay
        jsPsych.pluginAPI.setTimeout(function() { 
            prompt.innerHTML = '<p>Player 2 made a pick! &#128526;</p>'
        }, player_2_timer)
      },
      on_finish: function() {
        // clear the timeout at the end of the trial, in case a response was made and the trial ends
        // before the timeout duration is reached
        jsPsych.pluginAPI.clearAllTimeouts();
      },
      response_ends_trial: function(){
        if (trial_duration > player_2_timer){ true } else { false }
      }
    };
```

**Reply by becky-gilbert** [reply]
@petyaracz I think the problem is that, if `response_ends_trial` is ever `true`, then the participant's response will immediately end the trial, regardless of whether or not the player 2 timer has fired and the "Player 2 made a pick!" message is shown. Is that what you want to happen? Or do you want the trial to always wait at least as long as the player 2 timer, and then either end (if the participant has already responded at that point) or wait for the participant's response?

If it's the latter case, then I think what you're looking for is a 'minimum duration' parameter, which is something that other users have requested by that we haven't implemented yet. In this case you might be better off editing the plugin. You could add trial parameters for the player 2 prompt and timer duration, and edit the logic for when the trial should end. 

Let us know if you want any implementation guidance, and if you figure out a solution.

**Reply by petyaracz** [reply]
Thanks! I will take a stab at editing the plugin. Will check back here or on a more appropriate thread if I get stuck in a tangible way.

**Reply by petyaracz** [reply]
this __works__:

in `plugin-html-button-response.js`, change this:

```
if (trial.trial_duration !== null) {
              this.jsPsych.pluginAPI.setTimeout(end_trial, trial.trial_duration);
          }
```
to
```
if (trial.trial_duration !== null) {
              this.jsPsych.pluginAPI.setTimeout(check_trial, trial.trial_duration);
          }
```
and
```
if (trial.response_ends_trial) {
                  end_trial();
              }
```
to
```
if (trial.response_ends_trial || check_trial_called == true) {
                  end_trial();
              }
```
where
```
var check_trial = () => {
              check_trial_called = true
              if (response.button !== null) {
                console.log('called: check_trial, got response');
                end_trial();
              }
              console.log('called: check_trial, no response yet');
          }
```
and
```
var check_trial_called = false
```
then
```
var trial_ends_on_timer = {
      type: jsPsychHtmlButtonResponseMod,
      stimulus: 'Click the button.',
      choices: ['Ok'],
      trial_duration: 5000,
      response_ends_trial: false
    };
```
only ends trial if both button clicked and time elapsed.

modified plugin code [here](https://gist.github.com/petyaracz/8157301fe58981c93efb3f95de184242).

I might try and add the prompt change thingie and submit this as a plugin (though rn variable assignment is all over the place).

**Reply by becky-gilbert** [reply]
Great, thanks!

**Comment by petyaracz** [comment]
One alternative is to break up the trial and pretend to check for player 2 after the player picked. 

So, sometimes (here, 50% of the time) player 2 already made up their mind, sometimes they are still thinking. If they are still thinking, it can take a random amount of time (here, 2s).

The problem with this is that how long player 2 takes is now "how long the player takes + some time maybe" meaning that if a participant takes ages they might start to suspect that there is a reason why the purported player 2 seems roughly as slow as them, or even slower. I might be overthinking this though.

```
<!DOCTYPE html>
<html>
  <head>
    <title>Fake dyad experiment</title>
    <script src="https://unpkg.com/jspsych@7.0.0"></script>
    <script src="https://unpkg.com/@jspsych/plugin-html-button-response@1.0.0"></script>
    <link href="https://unpkg.com/jspsych@7.0.0/css/jspsych.css" rel="stylesheet" type="text/css" />
    <meta charset="UTF-8">
  </head>
  <body></body>
  <script>
    const jsPsych = initJsPsych();

    var trial = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'The name of a popular red fruit is:',
      choices: ['apple','carburator']
    };

    var player2_1 = {
        type: jsPsychHtmlButtonResponse,
        stimulus: 'Checking player 2...',
        choices: [],
        trial_duration: 1000
    }

    var player2_2 = {
        type: jsPsychHtmlButtonResponse,
        stimulus: 'Player 2 thinking...',
        choices: [],
        trial_duration: 2000
    }

    var player2_3 = {
        type: jsPsychHtmlButtonResponse,
        stimulus: 'Player 2 picked.',
        choices: [],
        trial_duration: 750
    }

    var if_node = {
        timeline: [player2_2],
        conditional_function: function(){
            var coin_flip = jsPsych.randomization.repeat([0,1], 1)[0];
            if(coin_flip == 0){
                return false;
            } else {
                return true;
            }
        }
    }

    var final = {
      type: jsPsychHtmlButtonResponse,
      stimulus: 'Thanks, bye.',
      choices: ['Ok']
    };

    jsPsych.run([trial,player2_1,if_node,player2_3,final]);
  </script>
</html>
```
